### ISSUE TITLE
feat: read `vim.opt.listchars` for the default config

### ISSUE THREAD

**Issue by vollowx** [mcauley-penney/visual-whitespace.nvim#5]
Read the content from `:help 'listchars'` for the default config, which reduces duplicated icon config.

**Comment by mcauley-penney** [comment]
I'm open to this, and have something really basic already implemented for it, but I'll have to think about it a bit more. I don't want to force users into it so it would definitely be opt-in.
 
It is simple for `eol` but there are multiple types of space and tab listchars.

To adhere to listchars strictly, I'll have to check how many spaces there are and where they are at in the line, given that listchars affords `space`, `lead`, `trail`, `multispace`, `leadmultispace`.

If the file uses tabs, the plugin will have to deal with the fact that tabs and the tab listchars have flexible lengths, such as with `tab:xy` and `tab:xyz`. I'll have to check at least the tabstop length, get the tab listchar, then trim it correctly (it is shortened and expanded differently depending on if you give `xy` or `xyz`). 

If we expect strict adherence to listchars and you have listchars set but don't use each of the ones listed above, it becomes worthless because you have to set options *somewhere* anyway. If you do use each of the given listchars, you can already see them in visual mode. If we do not expect strict adherence, we then have to choose which listchars to use.

**Comment by vollowx** [comment]
Thanks for your reply first!

To be honest, I didn't think of the complexity of reading `vim.opt.listchars` as default config before, as I personally use single-char for each type. Now configurating separated seems to be the better choice.

However, there's another idea: we could only read the listchars if its length is 1, otherwise we just ignore it.

**Comment by mcauley-penney** [comment]
I'm coming back to this idea after a long time. I think that `listchars` is antithetical to this plugin. If `list` is set, we don't really need this plugin. If list is not set, the user is probably not setting `listchars`. If they are setting that option only for this plugin, they shouldn't. It makes no sense to set builtin options to configure a plugin, I think.

~~What do you think about this? I am leaning towards closing this issue.~~ Closing

**Comment by bew** [comment]
> It makes no sense to set builtin options to configure a plugin, I think.

I disagree on principle, if a plugin can enhance builtin features without introducing new options it should 🤔
This also makes sure there is no duplication between the plugin's configuration & builtin options when the plugin is enabled or not

**Comment by mcauley-penney** [comment]
> no duplication between the plugin's configuration & builtin options

This seems easily solvable by defining a table with relevant characters somewhere and assigning it to both `listchars` and this plugin's option. I get that it means having an extra table and some `require` statements, but adding functionality to the plugin to avoid that seems excessive.

> if a plugin can enhance builtin features without introducing new options it should

I can see the merit of this at the same time that there is some divergence between what this plugin affords and what Neovim already has. [As I mention above](https://github.com/mcauley-penney/visual-whitespace.nvim/issues/5#issuecomment-2027828532), there are a lot of types of white space to check for and I don't know if I will support these, and Neovim doesn't support all of the `eol` types that this plugin does. Basically, there is only some overlap between `listchars` and what this plugin does, specifically for `space` and `nbsp`. 

I think that it would consequently be confusing to have inconsistent settings appearing from `listchars`. For example, I can foresee having to explain in the docs that the plugin will automatically use your listchar options for space and nbsp, but 

- not eol because listchars only affords one type of eol where this plugin affords one for each fileformat type, and
- not tab, because Vim/Neovim has a more advanced way of rendering tabs, allowing them to have expanding tabs where this plugin only places one character per tab

If I want to be more faithful to Neovim, I have to include functionality to deal with tabs, also mentioned above. Even then, the point about eol chars remains. So (and correct me if I'm wrong) this issue essentially becomes "Please support variable-length tab rendering". Is that desirable?

Edit: As an amendment to this, I have never intended on supporting all of the white space types that Neovim does. My interest has been in duplicating the functionality of VSCode. For example, I haven't been planning on supporting any types of spaces other than just a single space character. It isn't exactly true that I'm enhancing built-in features, in my opinion.

**Comment by vollowx** [comment]
> If list is set, we don't really need this plugin. If list is not set, the user is probably not setting listchars. If they are setting that option only for this plugin, they shouldn't.

That makes sense. However, my original thought is that we could sync the look of this plugin with the native look by default, which have nothing to do with the `list` option. Though I'm fine with setting this option manually.

**Comment by mcauley-penney** [comment]
I'm currently implementing a way for the user to choose which white space types they show. It's not done yet, but this may be a feature soon
